### ISSUE TITLE
Micropayments incentivized bridges

### ISSUE THREAD

**Issue by ple1n** [oxen-io/lokinet#2006]
Various ways of censorship-evasion have been tried, but I haven't seen an actually useful tokenomics-based project. Existing ones are merely VPNs, and some of them don't even support Linux.

Bridge providers should be incentivized to tweak their protocol parameters or offer their own protocols. The recommended ones would be obfs4, and its variant, widely used by Chinese right now, [sosistab](https://github.com/geph-official/sosistab), or meek which exploits censors' whitelist. There are three ways of censorship-circumvention, obfuscation, mimicry (parrot legit protocols), and tunneling (includes domain fronting and video streams). If bridge providers are to invent their own protocols, how shall they ship the clients. maybe running them in a wasm container.

Is it possible to construct a micropayment scheme on Oxen ? Orchid protocol has a probabilistic nanopayment system, but it requires smart contracts. Should the micropayment be made chain-agnostic ? The users of bridges tend to have anonymity need too.

**Comment by majestrate** [comment]
> Is it possible to construct a micropayment scheme on Oxen ? 

on chain payment logic with oxen is the next major feature we want to add to lokinet going forward. right now it is primarily meant to be used to facilitate service discovery and payments with 1 hop snapps on lokinet (exit market). i would like to figure out how to do something similar for network bridging infra. adding that on chain logic to bridging may be slightly more involved vs the exit market but i dont know if it is or by how much, just a gut feeling.

**Comment by ple1n** [comment]
> exit market

I think I read about paid exits somewhere. Is there docs on exit node payment ? Are they paid directly on Oxen layer one, which is spammy if doing micropayments ?

Btw, maybe wOxen should be deployed on eth layer 2 to lower gas fees.

**Comment by majestrate** [comment]
> > exit market
> 
> I think I read about paid exits somewhere. Is there docs on exit node payment ? Are they paid directly on Oxen layer one, which is spammy if doing micropayments ?

we would want it to be using oxen as that allows us to do really cool things like zk proofs to exits themselves without the exit knowing who is authenticating, just that they paid for transit. microtx arent really needed, as those can be negotiated via offchain side channel (like lokinet) and then have the aggregate published on chain.

> 
> Btw, maybe wOxen should be deployed on eth layer 2 to lower gas fees.

i do not see any compelling reasons to involve eth.

**Comment by ple1n** [comment]
> i do not see any compelling reasons to involve eth.

It's listed oxen.io https://ethereum.oxen.io/. It's not to involve. The token bridge can increase liquidity. Not all people can use centralized exchanges to deposit funds. 

also, censors probe IPs for bridges, so bridge information is a scarce resource

microtx allows for more fine-grained control. changing bridges should be often

**Comment by majestrate** [comment]
> > i do not see any compelling reasons to involve eth.
> 
> It's listed oxen.io https://ethereum.oxen.io/. It's not to involve. The token bridge can increase liquidity. Not all people can use centralized exchanges to deposit funds. 
> 
this looks like it is a problem addressed by chainflip and is at best an implementation detail that should remain far out of scope from the protocol designs in lokinet.

> also, censors probe IPs for bridges, so bridge information is a scarce resource
> 
yes it is, and this is much more relevent to the design of incentivised bridging.

> microtx allows for more fine-grained control. changing bridges should be often

i dont think it is a scalable way to do it, you would be better off negotiating in a side channel with all of the bridges you plan on using and transmit a net result. i dont think putting awareness of most of this kind of state into lokinet is desirable.

**Comment by nullchinchilla** [comment]
so I think the hardest part of bridge distribution is not protocol obfuscation: protocol obfuscation is naturally incentivized if users just pay bridges for traffic, since bridges are therefore incentivized to not b e blocked.

the hardest part is making sure different users automatically retrieve different bridges, over some unblockable channel. the way Geph, Snowflake, etc do it is *inherently* based on a centralized, non-transparent, permissioned rendezvous server.

**Comment by majestrate** [comment]
> so I think the hardest part of bridge distribution is not protocol obfuscation: protocol obfuscation is naturally incentivized if users just pay bridges for traffic, since bridges are therefore incentivized to not b e blocked.
> 

This is my assessment on it too, the incentives are all about wide accessibility. more ASN diversity more better.

> the hardest part is making sure different users automatically retrieve different bridges, over some unblockable channel. the way Geph, Snowflake, etc do it is _inherently_ based on a centralized, non-transparent, permissioned rendezvous server.

i dont think the normal network bridging setup will work out after seeing the pain tor has gone through over the past almost decade now. the idea i have in my head for our net censor mitigation is to do a kind of recursive bridging, basically a full of mesh net which i capable of transiting over the l3 or l2 of your choosing. routing around the DPI by using an alternative l2 looks to be really the only way to truly getting out of networks like chinanet going forward.

as for the incentives, maybe if there was a way to prove network diversity (with zero knoledgeness) given your set of transit bridges you offer in a future theoretical setup, we could structure incentives such that it promotes a wide diversity of ASN.

**Comment by ple1n** [comment]
> basically a full of mesh net which i capable of transiting over the l3 or l2 of your choosing

wdym by l2, l3 ? yggdrasil ?

> making sure different users automatically retrieve different bridges

so, the censor is collecting bridges through the faucet. it reminds me of proof of personhood. even PoP isn't enough, because censors can always hire/coerce/bribe people to pass the test. and censors may buy up bridges and block them.

maybe https://appliedzkp.org could be of use. users may anonymously import reputation from sources.

**Comment by majestrate** [comment]
> > basically a full of mesh net which i capable of transiting over the l3 or l2 of your choosing
> 
> wdym by l2, l3 ? yggdrasil ?

yes, i have been planning on yoinking thier dht spanning tree protocol, pondering on if i'd do a multi root version where each root is the snode you eventually use.


> 
> > making sure different users automatically retrieve different bridges
> 
> so, the censor is collecting bridges through the faucet. it reminds me of proof of personhood. even PoP isn't enough, because censors can always hire/coerce/bribe people to pass the test. and censors may buy up bridges and block them.
> 

i feel like bridges have to be manually out of band. this does not exclude physical proximity autodiscovery tho. i do ponder the defaults in that case, if we should have that be opt out or opt in.

> maybe https://appliedzkp.org could be of use. users may anonymously import reputation from sources.
> 
> 

i'll peek at it

**Comment by ple1n** [comment]
It may serve as an alternative requirement for running a relay node. Some people [criticize](https://anonymousplanet.org/guide#appendix-b7-caution-about-session-messenger) lokinet for having a high entry barrier. If you want non-colluding nodes, zk-rep should work as a stake too (voluntary relays without rewards). 

Many censor-evading services in China are two-hoped. They relay traffic in China because of better routes and in different regions the strictness of censorship differs. Some people relay their traffic inside China to prevent their university or local admins from suspecting them. (constantly connecting a small set of foreign IPs in obscure protocols is sus). Running relays in China is risky tho.

**Comment by majestrate** [comment]
> It may serve as an alternative requirement for running a relay node. Some people [criticize](https://anonymousplanet.org/guide#appendix-b7-caution-about-session-messenger) lokinet for having a high entry barrier. If you want non-colluding nodes, zk-rep should work as a stake too (voluntary relays without rewards). 
> 

that blog author in particular comes off as intellectually dishonest tbh. in the past, ( i think it was them) they complained about how no one should use lokinet because "it doesn't compile on rpi and you are forced to use binaries and thus it may as well be closed source thus it's backdoored don't trust it" or something like that, which is provable false and easily refutable as all our ci for arm runs on native rpi, even the debs are built on them. then they changed it to complain about how we don't support openbsd (instead of rpi) so therefore yada yada yada same pile of crap goes here. we never actually had support for obsd either so idk why this is even an issue to pipe on about especially given what it used to say.

just from that history of having that kind of insufferably lazy non excuses thrown at us i am very reluctant to take such banal gripes at face value anymore. it always comes from the same clique that does not contribute anything nor do they seem willing to communicate or collaborate in a way that isn't fodder for fud. 

it's just so tiresome tbh.

> Many censor-evading services in China are two-hoped. They relay traffic in China because of better routes and in different regions the strictness of censorship differs. Some people relay their traffic inside China to prevent their university or local admins from suspecting them. (constantly connecting a small set of foreign IPs in obscure protocols is sus). Running relays in China is risky tho. 

this is why our theoretical bridge protocol is "recursive".

**Comment by majestrate** [comment]
> > Is it possible to construct a micropayment scheme on Oxen ? 
> 
> on chain payment logic with oxen is the next major feature we want to add to lokinet going forward. right now it is primarily meant to be used to facilitate service discovery and payments with 1 hop snapps on lokinet (exit market). i would like to figure out how to do something similar for network bridging infra. adding that on chain logic to bridging may be slightly more involved vs the exit market but i dont know if it is or by how much, just a gut feeling.

also keep in mind when you add a huge paragraph into your original post 6 months after you made it in such a way it looks like i cherry picked a reply, it's not a good look on you.

**Comment by ple1n** [comment]
🫠i tend to edit that comment so it is like a summary of the thread. the owner or someone muted me in their room for one minute, bcs i said your product doesn't need compatibility with tor (i have no idea what 'tor compatibility' is and he refused to elaborate)

in his matrix room
> It was designed to be intentionally incompatible with Tor as well.

I do agree that the staking requirement is somewhat preventing potential honest nodes from entering the network.

**Comment by ghost** [comment]
> I do agree that the staking requirement is somewhat preventing potential honest nodes from entering the network.

I agree as well and I proposed an upper-bound-lower-bound staking requirement in https://github.com/oxen-io/oxen-core/issues/1635
It's quite a challenge task in lots of design and implementation details though.

**Comment by majestrate** [comment]
> 🫠i tend to edit that comment so it is like a summary of the thread. 

that's makes much more sense. keep in mind i am a bit stressed so i can make mistakes, please be patient with me.

> the owner or someone muted me in their room for one minute, bcs i said your product doesn't need compatibility with tor (i have no idea what 'tor compatibility' is and he refused to elaborate)
> 
> in his matrix room

people are allowed to have their opinions, over all i dont think we need their (or really anyone's) formal certification of goodness. i am not a fan of the "trust the experts" routine as it trains people to be dependent on a third party to form opinions on what they should trust, what software they should use or how to live their life. 

we can be a better project and have a better network without such wantings.

> 
> > It was designed to be intentionally incompatible with Tor as well.
> 
> I do agree that the staking requirement is somewhat preventing potential honest nodes from entering the network.

It does filter out the low effort nodes much better than potentially honest nodes as the ones that get in are not "potentially honest" but they are committing to be honest with a large chunk of skin in the game.

**Comment by majestrate** [comment]
> > I do agree that the staking requirement is somewhat preventing potential honest nodes from entering the network.
> 
> I agree as well and I proposed an upper-bound-lower-bound staking requirement in [oxen-io/oxen-core#1635](https://github.com/oxen-io/oxen-core/issues/1635) It's quite a challenge task in lots of design and implantation details though.

yes, it's a major change. it looks like a very insightful observation and suggestion but man will it take time to properly sus out a direction forward on that.

also why the hell did it give me edit access on your post i didnt want to do that. i wanted to click reply. jfc github STOP.
